macro/R/continuous_step.R

```diff
@@ -1,51 +1,333 @@
+
+  check_continuous_setup(individuals, internal_transitions)
+
+  list(
+    state = individuals,
+    transitions = internal_transitions,
+    variables = variables,
+    observer = observer,
+    time = 0,
+    trajectory = vector(mode = "list", length = 100L),
+    trajectory_cnt = 0L
+    )
+}
+
+
+#' Given a simulation object, calculate enabled transitions.
+#'
+#' @param simulation This object is created by \code{\link{continuous_simulation}}
+#' @export
+init_simulation <- function(simulation) {
+  transitions <- simulation$transitions
+  # The current time and next firing times are part of the next state of the system.
+  # If a firing time is Inf, that means it isn't scheduled.
+  simulation$state <- initialize_times(simulation$state, transitions)
+  simulation
+}
+
+
+next_step_over_time <- function(duration) {
+  function(state, step_cnt, current_time, next_time) {
+    next_time >= duration
+  }
+}
+
+
+#' Observe each transition.
+#'
+#' The memory use of this method could be improved. We could construct a
+#' data.table into which to store the trajectory, so that it overwrites
+#' lines in the data.table. By returning lists, we're churning memory.
+#'
+#' You may want to augment this to record the id of the individual.
+#'
+#' @param transition_name The string name of the transition.
+#' @param former_state a list describing the individual's state before the transition
+#' @param new_state a list describing the individual's state after the transition
+#' @param time The time at which this transition fires.
+#' @return a list with the name and time.
+#' @export
+observe_continuous <- function(transition_name, former_state, new_state, time) {
+  list(name = transition_name, time = time)
+}
+
+
+#' Run the simulation for a certain time.
+#'
+#' @param simulation This object is created by \code{\link{continuous_simulation}}
+#' @param duration How long to run
+#' @return The trajectory of the run.
+#' @export
+run_simulation <- function(simulation, duration) {
+  step_idx <- 0L
+  current_time <- 0
+  stop_condition <- next_step_over_time(duration)
+  observer <- observe_continuous
+  individuals <- simulation$state
+  if ("trajectory" %in% names(simulation)) {
+    trajectory <- simulation$trajectory
+  } else {
+    trajectory <- vector(mode = "list", length = 256L)
+  }
+  # Use the count of trajectory entries in case someone calls this without
+  # clearing the trajectory. This lets us tack onto the end of the trajectory.
+  if ("trajectory_cnt" %in% names(simulation)) {
+    trajectory_cnt <- simulation$trajectory_cnt
+  } else {
+    trajectory_cnt <- 0
+  }
+  while (TRUE) {
+    individual <- as.list(individuals[order(when)][1])
+    should_end <- stop_condition(individuals, step_idx, current_time, individual$when)
+    if (is.infinite(individual$when) || should_end) {
+      break
+    }
+    new_state <- update_individual(individual, simulation$transitions, observer)
+    individuals[order(when)][1] <- new_state$individual
```

Comment:
same issue, replace LHS with `individuals[order(individuals$when)[1],]`

Reply:
That makes sense. I'll change this and the one above. I like this comma because it matches the data.frame version, so it's more familiar.